### PR TITLE
SVG Parser: path "d" attributes

### PR DESCRIPTION
**Pull request by pengmai**
As of these changes, the parser is able to successfully parse the dynamically generated sample graph. A/B Testing was performed via database dumps with all of the static graphs to ensure that the resulting path values were identical with the new path parser.

Additionally, one of the static graphs contained a single value in scientific notation, so this PR includes support for parsing those values.

### CODE REVIEW

File: app/Svg/Parser.hs [hunk 1]
```diff
@@ -344,82 +367,106 @@ getTransform :: Tag T.Text -> Point
...
--- | Parses a translation String into a tuple of Float.
+-- | Parses a translation String into a tuple of Float, ignoring scaling and
+-- rotation.
 parseTransform :: T.Text -> Point
 parseTransform "" = (0, 0)
 parseTransform transform =
     parseVal parser transform
     where
-        parser = do
+        parser = P.many (scale <|> rotate) >> translate
```
Comment (david-yz-liu) [inline app/Svg/Parser.hs]: Use `P.sepBy` to move the `P.spaces` out of the definition of scale/rotate.

File: app/Svg/Parser.hs [hunk 2]
```diff
@@ -344,82 +367,106 @@ getTransform :: Tag T.Text -> Point
...
-      convertToPoint (x, y) =
-        case (readMaybe (T.unpack x), readMaybe (T.unpack y)) of
-            (Just m, Just n) -> Just (m, n)
-            _ -> Nothing
+        -- Keep track of the (first point, path mode, most recent point)
+        initialState = PathDState { firstPoint = (0, 0)
+                                  , mode = AbsoluteMove
+                                  , currentPoint = (0, 0) }
+
+        parser = do
```
Comment (david-yz-liu) [inline app/Svg/Parser.hs]: Include type signatures for these helpers.

File: app/Svg/Parser.hs [hunk 3]
```diff
@@ -344,82 +367,106 @@ getTransform :: Tag T.Text -> Point
...
+            p <- parseStep
+            -- Record the first point added.
+            _ <- P.modifyState $ \st -> st { firstPoint = p, currentPoint = p }
+            rest <- P.many parseStep
+            return (p : rest)
+        parseStep = do
+            _ <- (P.option () $ bezier <|>
+                absoluteMove <|> relativeMove <|>
+                absHorizontalMove <|> absVerticalMove <|>
+                relHorizontalMove <|> relVerticalMove) >> P.spaces
```
Comment (david-yz-liu) [inline app/Svg/Parser.hs]: In general, put the `P.spaces` on a new line (`_ <- P.spaces`).

File: app/Svg/Parser.hs [hunk 4]
```diff
@@ -344,82 +367,106 @@ getTransform :: Tag T.Text -> Point
...
+                                  , currentPoint = (0, 0) }
+
+        parser = do
+            p <- parseStep
+            -- Record the first point added.
+            _ <- P.modifyState $ \st -> st { firstPoint = p, currentPoint = p }
+            rest <- P.many parseStep
+            return (p : rest)
+        parseStep = do
+            _ <- (P.option () $ bezier <|>
```
Comment (david-yz-liu) [inline app/Svg/Parser.hs]: Use `P.choice` to simplify the `<|>`s.

File: app/Svg/Parser.hs [hunk 5]
```diff
@@ -344,82 +367,106 @@ getTransform :: Tag T.Text -> Point
...
+                absoluteMove <|> relativeMove <|>
+                absHorizontalMove <|> absVerticalMove <|>
+                relHorizontalMove <|> relVerticalMove) >> P.spaces
+            p <- closeLoop <|> moveArg
+            state <- P.getState
+            let (currentX, currentY) = currentPoint state
```
Comment (david-yz-liu) [inline app/Svg/Parser.hs]: I'd like you to move this part into a separate helper of type `Point -> PathDState -> PathDState ` (extract the relevant parts of the current state in the helper). Then you can do `let p' = (yourHelper p)` etc. in the main parser. Also see my above comment about `P.spaces`.

File: app/Svg/Parser.hs [hunk 6]
```diff
@@ -26,13 +26,18 @@ import Database.DataType
...
 import Text.Read (readMaybe)
 import Data.Char (isSpace)
 import qualified Data.Text as T
 import Data.Text.IO as T (readFile)
 import Data.List.Split (splitOn)
+import Data.Functor.Identity (Identity)
 
+-- A Parsec parser with user defined state. This can be made stateless by
+-- assigning () to u.
+type StatefulParser u a = P.ParsecT String u Identity a
```
Comment (david-yz-liu) [inline app/Svg/Parser.hs]: I see what you're trying to do here, but there's the `P.Parsec` type synonym already that we should just use. You can turn the `Parser a` uses below into `P.Parsec String u a`.

File: app/Svg/Parser.hs (hunk 3 above)
Comment (pengmai) [reply app/Svg/Parser.hs]: Right, that was an attempt to save lines that resulted in making code less readable in a bunch of cases. Does that also apply to instances like `bezier`, where everything ends up getting discarded?

**Pull request merged**